content/blog/zephyr-tracing-and-profiling/index.md

```diff
@@ -0,0 +1,279 @@
+TRACING_SEMAPHORE=y/n
+TRACING_HEAP=y/n
+TRACING_STACK=y/n
+TRACING_PM=y/n
+(...)
+```
+
+If a large amount of data is to be generated, this might be useful to limit tracing only to the objects of interest, both to save buffer and provide a cleaner data acquisition.
+
+Another interesting feature is the possibility of defining user functions to be called at tracing events, as *headers* shown below. In a debugging situation or when external equipment (e.g oscilloscope or data logger) is used, it can be useful to add custom code to the callbacks to drive a GPIO, for instance, in order to sync external acquisition data with the log provided by the *tracing* subsystem.
+
+<a href="img/img002.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img002.webp"
+        alt=""
+        caption="Trace user functions"
+    >}}
+</a>
+
+The following macros [[2]] are placed on Zephyr’s subsystems code to track important functions of the operating system:
+
+```
+SYS_PORT_TRACING_FUNC(type, func, …)
+SYS_PORT_TRACING_FUNC_ENTER(type, func, …)
+SYS_PORT_TRACING_FUNC_EXIT(type, func, ...)
+```
+
+This means that depending on the functionality being probed, the *tracing* subsystem might already have hooks that will acquire data by default, as can be seen in the example below:
+
+```c
+/* Called when data needs to be sent to network */
+int net_send_data(struct net_pkt *pkt)
+{
+    int status;
+    int ret;
+
+    SYS_PORT_TRACING_FUNC_ENTER(net, send_data, pkt);
+
+    (...)
+}
+```
+
+Instrumenting application, however, is currently not supported without manually implementing some code inside the *tracing* subsystem. For the time being, there are 3rd party tools that support such feature, and a new tool is expected to be integrated to Zephyr in the future that might as well support it [[3]].
+
+## Zephyr’s profiling tool
+
+The current profiling implementation monitors the function stack in order to allow examining the software execution at given intervals. To enable it, the following configs must be defined:
+
+```
+CONFIG_PROFILING=y
+CONFIG_PROFILING_PERF=y
+CONFIG_PROFILING_PERF_BUFFER_SIZE=2048
+CONFIG_THREAD_STACK_INFO=y
+CONFIG_SMP=n
+CONFIG_SHELL=y
+CONFIG_FRAME_POINTER=y
+```
+
+Using these directives it is possible to obtain a log that can be processed by tools and yield a composed view of what the system is executing:
+
+<a href="img/img003.webp" target="_blank">
+    {{< figure
+        default=true
+        src="img/img003.webp"
+        alt=""
+        caption="Profiling sample graph"
+    >}}
+</a>
+
+The graph above shows a timeline of the functions pushed onto the stack, including all sub-function calls. While monitoring the function stack can be useful for analyzing certain kinds of problems, we won’t discuss this feature in detail, as it offers limited functionality for now. As mentioned earlier, a new subsystem is being developed that may provide better functionality [[3]].
+
+## Percepio’s Tracealyzer

```

Comment:
Please review all `'s` used in the documentation. In the case above, `Percepio´s` should be `Percepio's`. There are other places like `Zephyr´s` instead of `Zephyr's`.